### PR TITLE
608 bug itchio deployment fails with missing production salt crash due to GitHub actions cache

### LINKED ISSUE #608: [BUG] Itch.io deployment fails with "Missing production salt" crash due to GitHub Actions cache
**Title**: [BUG] Itch.io deployment fails with "Missing production salt" crash due to GitHub Actions cache

**Godot Version**: 4.5 stable

**OS**: Web Export (via Ubuntu-latest GitHub Actions runner)

**Steps to Reproduce**:

1. Trigger the `.github/workflows/deploy_to_itch.yml` deployment pipeline.
2. The `inject_salt.sh` script successfully replaces the `"CI_INJECT_SALT_HERE"` placeholder in `scripts/core/globals.gd` with the real production secret.
3. The `firebelley/godot-export` action executes with the `cache: true` parameter.
4. Open the newly deployed build on itch.io and inspect the browser developer console.

**Expected Behavior**:

The game should detect the successfully injected production salt, generate a secure SHA-256 encryption key, and boot normally to the main menu.

**Actual Behavior**:

The game crashes instantly on startup. The production security guard triggers because the compiled game still contains the `"CI_INJECT_SALT_HERE"` placeholder, completely ignoring the bash injection step.

The browser console outputs:

```text
[DEBUG] ⚙️ ENV CHECK (is_automated_test): false
ERROR: CRITICAL SECURITY ERROR: Missing production salt.
  at: push_error (core/variant/variant_utility.cpp:1024)
ERROR: CRITICAL SECURITY ERROR: Missing production salt.
  at: crash (core/core_bind.cpp:336)
Error starting Godot: RuntimeError: unreachable

```

**Screenshots**:

Image

**Additional Context**:

This is a collision between our custom CI pipeline and the Godot export cache.

Because `deploy_to_itch.yml` uses `cache: true` for the `godot-export` action, it pulls down a `.godot/` cache from a previous run. Even though our `inject_salt.sh` script successfully mutates the `globals.gd` source file, the Godot engine relies on the cached compiled bytecode during the headless export phase. As a result, the old placeholder string is packaged into the final `.pck` file.

**Proposed Fix**:

Update `.github/workflows/deploy_to_itch.yml` to explicitly disable the export cache so Godot is forced to recompile the freshly injected GDScript:

```yaml
      - name: "Export Godot to Web"
        id: "export"
        uses: "firebelley/godot-export@930577654862a320eef793f399ee911b4479efb9"
        with:
          # ...
          cache: false  # <-- Change this from true to false
```

### PR DESCRIPTION
**Pull request by ikostan**
---
name: Default Pull Request Template
about: Suggesting changes to SkyLockAssault
title: ''
labels: ''
assignees: ''
---

Here is a detailed summary of everything we have diagnosed and fixed so far, perfectly formatted so you can drop it directly into Issue #608 to document the trail.

---

## Bug Summary: Production Deployment Fails with "Missing production salt"

**The Core Issue:**

The itch.io deployment crashed immediately on boot with `ERROR: CRITICAL SECURITY ERROR: Missing production salt.` The compiled Godot export still contained the literal `"CI_INJECT_SALT_HERE"` placeholder, meaning the bash injection step failed to permanently modify the bytecode used in the final `.pck`.

**Why tests didn't catch it:**

Automated Playwright tests and local simulations run with the `ci` feature flag injected (`OS.has_feature("ci")`). This flag explicitly bypasses the production security crash block in `globals.gd`. Therefore, if the injection silently failed in testing, the game simply used the placeholder as the key and booted normally. In production, without the `ci` flag, the failsafe correctly detected the placeholder and crashed the game.

### Applied Fixes & Pipeline Hardening

**1. Disabled the Godot Export Cache (`deploy_to_itch.yml`)**

* **The Risk:** The `firebelley/godot-export` action was running with `cache: true`. This meant Godot could pull pre-compiled `.godot/` cache files from previous runs that were compiled *before* the current salt injection step.
* **The Fix:** Set `cache: false` to force Godot to recompile `globals.gd` from scratch, guaranteeing the newly injected script is read.

**2. Fixed the Butler Push Target (`deploy_to_itch.yml`)**

* **The Risk:** The deployment step was using `${{ steps.export.outputs.archive_directory }}/Web.zip`, which points to the raw Godot export output. This completely bypassed the custom `patch_index_js.sh` step and the re-zipped archive.
* **The Fix:** Hardcoded the butler target to explicitly push the post-processed `export/Web.zip`.

**3. Hardened the Injection Script (`inject_salt.sh`)**

* **Strict Error Handling:** Added `set -e` to the top of the script so it immediately halts the GitHub Action if `sed` or any other command fails.
* **Newline Stripping:** Added `tr -d '\r\n'` to the `PRODUCTION_SALT` assignment to strip invisible trailing newlines or carriage returns that are often accidentally copied into GitHub Secrets, preventing silent `sed` syntax errors.
* **Explicit Verification:** Added a final `grep` check at the bottom of the script. It actively scans the file after injection and forces a fatal exit code `1` if the `"CI_INJECT_SALT_HERE"` placeholder is still found. This guarantees silent failures can never reach the export step again.

**4. Rotated & Sanitized the Secret**

* **The Fix:** Updated the `PRODUCTION_SALT` in GitHub Repository Secrets to a highly entropic, pipeline-safe alphanumeric string (Base64 format). This eliminates any risk of complex characters (`'`, `"`, `\`, `$`) inadvertently breaking the bash parser during the runner execution.

---

## Related Issue

Closes #ISSUE_NUMBER (if applicable)

## Changes

- [ ] List key changes here (e.g., "Updated Jump.gd to use Godot 4.4's new Tween
  system")
- [ ] Any breaking changes? (e.g., "Deprecated old signal; migrate to new one")

## Testing

- [ ] Ran the game in Godot v4.5 editor—describe what you tested (e.g., "Jump
  works on Win10 with 60 FPS")
- [ ] Any new unit tests added? (Link to test scene if yes)
- [ ] Screenshots/GIFs if UI-related: (Attach below)

## Checklist

- [ ] Code follows Godot style guide (e.g., snake_case for variables)
- [ ] No console errors in editor/output
- [ ] Ready for review!

## Additional Notes

Anything else? (e.g., "Tested on Win10 64-bit; needs Linux validation")

## Summary by Sourcery

Harden salt injection for production builds and fix the itch.io deployment workflow to reliably ship a patched, verified Web export.

Bug Fixes:
- Prevent deploy script from crashing on missing PRODUCTION_SALT by safely handling unset environment variables and empty secrets.
- Ensure itch.io deployment always uses the patched Web.zip instead of mixing raw export files with the archive.

Enhancements:
- Strip newlines from multiline PRODUCTION_SALT values before injection while preserving meaningful content, and verify placeholders are fully replaced.

Build:
- Adjust the itch.io GitHub Actions workflow to unzip, patch, and re-zip the Web export in a dedicated directory tied to the export step output.

Tests:
- Expand CI tests for salt injection to cover multiline and whitespace-only secrets, as well as behavior when the secret environment variable is completely unset.

## Summary by Sourcery

Harden salt injection and deployment handling for web exports to prevent missing production salt failures and verify patched builds before itch.io deployment.

Bug Fixes:
- Prevent salt injection from proceeding with unset, empty, or newline-only PRODUCTION_SALT values and ensure the target file remains unchanged on failure.
- Ensure the final itch.io deployment always uses the patched Web.zip by re-zipping the processed export at the workflow’s configured archive path and verifying the placeholder is fully removed.

Enhancements:
- Support multiline and CRLF secrets by stripping newlines before injection while preserving meaningful whitespace, and log the generated encryption key length for easier debugging of production salt issues.

Build:
- Tighten the itch.io GitHub Actions workflow with strict shell error handling, clean staging for unzipping, patched re-zipping of web exports, and a final archive-level check that no salt placeholder remains.

Documentation:
- Document the improved handling of PRODUCTION_SALT and the updated itch.io deployment workflow in the milestone notes.

Tests:
- Extend CI tests to cover multiline, whitespace-only, and completely unset PRODUCTION_SALT values, as well as positive-path injection behavior after newline stripping.

## Summary by CodeRabbit

* **Chores**
  * Stricter CI/script execution and safer secret handling with validation, normalization, and post-injection verification; packaging flow now reliably re-zips the patched web export before deployment
  * Minor internal cleanup to key-generation handling

* **Tests**
  * CI tests expanded to cover unset, empty-after-stripping, and multiline secrets; verify stripping behavior and clear error messaging

* **Documentation**
  * Milestone notes updated to reflect the above changes

[Review Change Stack](https://app.coderabbit.ai/change-stack/ikostan/SkyLockAssault/pull/610?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: tests/ci/test_salt_injection.py [hunk 1]
```diff
@@ -252,3 +232,23 @@ def test_idempotency(repo_tmp):
...
+def test_injection_multiline_secret_stripped(repo_tmp):
+    """Validates that multiline secrets are safely stripped of newlines and injected, rather than failing."""
+    dummy_rel = f"{repo_tmp}/dummy_multiline.gd"
+    dummy_abs = Path(PROJECT_ROOT) / dummy_rel
+    original_content = 'var salt = "CI_INJECT_SALT_HERE"\n'
+
+    dummy_abs.write_text(original_content, encoding="utf-8")
+
+    # Pass a secret with a newline
+    result = run_injection(dummy_rel, "line1\nline2")
```
[the remark this answers is not in the document]
Comment (ikostan) [reply tests/ci/test_salt_injection.py]: While the test we wrote proved that the tr -d '\r\n' fix worked, the bot is right that using in is a bit loose. More importantly, since Godot development happens heavily on Windows, explicitly testing \r\n (CRLF) line endings is a fantastic idea to prevent future regressions.

Let's address all three of the bot's bullet points at once by converting that single test into a strictly-asserted, parameterized test.

**Comment by ikostan** [comment]
@sourcery-ai guide

File: tests/ci/test_salt_injection.py [hunk 2]
```diff
@@ -252,3 +232,23 @@ def test_idempotency(repo_tmp):
...
+
+    # It should now SUCCEED, because `tr -d '\r\n'` sanitizes the input
+    assert result.returncode == 0
+
+    content = dummy_abs.read_text(encoding="utf-8")
+
+    # The newline should be stripped completely from the final injected string
+    assert 'var salt = "line1line2"' in content
```
[the remark this answers is not in the document]
Comment (ikostan) [reply tests/ci/test_salt_injection.py]: CodeRabbit is complaining about the exact same thing Sourcery was whining about: the old version of the test only checked the Linux-style newline (`\n`).

Because we just replaced that single test with the robust `@pytest.mark.parametrize` block in the previous step, we've already covered it. If we look at the block we just added:

```python
    [
        ("line1\nline2", "line1line2"),  # Standard Linux/macOS LF
        ("line1\r\nline2", "line1line2"),  # <-- The Windows CRLF case CodeRabbit is asking for!
        ("\n\nline1\n\nline2\n\n", "line1line2"),  
        ("\r\n mixed \n newlines \r\n", " mixed  newlines "),  
    ]

```

We explicitly added the `"line1\r\nline2"` case. We are 100% covered for cross-platform Windows CRLF line endings.

**Comment by ikostan** [comment]
@sourcery-ai guide

**Comment by ikostan** [comment]
@sourcery-ai review

File: .github/scripts/inject_salt.sh [hunk 3]
```diff
@@ -13,22 +17,32 @@ if [ ! -f "$TARGET_FILE" ]; then
...
 
-# Cross-platform sed for in-place editing (macOS vs Linux)
 # Source the shared utilities
 source "$(dirname "$0")/ci_utils.sh"
 
 # 3. Replace the safe placeholder with the real secret
 sedi "s|\"CI_INJECT_SALT_HERE\"|\"$SED_ESCAPED\"|g" "$TARGET_FILE"
+
+# 🛑 4. EXPLICIT VERIFICATION: Ensure the placeholder was actually removed
+if grep -q "CI_INJECT_SALT_HERE" "$TARGET_FILE"; then
```
[the remark this answers is not in the document]
Comment (ikostan) [reply .github/scripts/inject_salt.sh]: Sourcery strikes again!

This is an incredibly pedantic suggestion from the bot, but technically, it is 100% correct bash scripting best practice.

[the remark this answers is not in the document]
Comment (ikostan) [reply .github/scripts/inject_salt.sh]: Did we manage to fix this?

File: .github/workflows/deploy_to_itch.yml [hunk 4]
```diff
@@ -64,15 +64,20 @@ jobs:
...
-        run: bash ./.github/scripts/patch_index_js.sh "export/web"
+        run: bash ./.github/scripts/patch_index_js.sh "export/patched_build"
 
       - name: "Re-zip Flattened Export"
         run: |
-          cd export/web
-          zip -r ../Web.zip .
+          # 1. Delete the original, unpatched zip
+          rm ${{ steps.export.outputs.archive_directory }}/Web.zip
+          # 2. Go into our patched folder
```
[the remark this answers is not in the document]
Comment (ikostan) [reply .github/workflows/deploy_to_itch.yml]: Sourcery strikes yet again!

As annoying as it is to have a bot nitpick every single line of your YAML, it is technically correct here. This is a classic Bash scripting gotcha.

If the GitHub Action runner ever decides to output your export to a directory that contains a space (for example, export/web build/), the command rm export/web build/Web.zip would break into two separate commands. Bash would try to delete a folder named export/web and a file named build/Web.zip, completely crashing your pipeline.

Wrapping the dynamic variable in double quotes ("...") forces Bash to treat the entire string as a single, unbreakable file path.

File: .github/scripts/inject_salt.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (ikostan) [reply .github/scripts/inject_salt.sh]: Can you check it again?

File: .github/scripts/inject_salt.sh [hunk 5]
```diff
@@ -13,22 +17,32 @@ if [ ! -f "$TARGET_FILE" ]; then
...
 
 # 2. Escape for sed replacement string
 SED_ESCAPED=$(printf '%s' "$GODOT_ESCAPED" | sed 's/\\/\\\\/g; s/&/\\&/g; s/|/\\|/g')
 
-# Cross-platform sed for in-place editing (macOS vs Linux)
 # Source the shared utilities
 source "$(dirname "$0")/ci_utils.sh"
 
 # 3. Replace the safe placeholder with the real secret
 sedi "s|\"CI_INJECT_SALT_HERE\"|\"$SED_ESCAPED\"|g" "$TARGET_FILE"
```
[the remark this answers is not in the document]
Comment (ikostan) [reply .github/scripts/inject_salt.sh:46]: Sourcery is completely hallucinating. It is suffering from a classic LLM misunderstanding of how Bash parsing and execution order actually works. We do not need to change a single line of code.

File: tests/ci/test_salt_injection.py [hunk 6]
```diff
@@ -252,3 +232,36 @@ def test_idempotency(repo_tmp):
...
+@pytest.mark.parametrize(
+    "secret_input, expected_injected",
+    [
+        ("line1\nline2", "line1line2"),  # Standard Linux/macOS LF
+        ("line1\r\nline2", "line1line2"),  # Windows CRLF
+        ("\n\nline1\n\nline2\n\n", "line1line2"),  # Multiple leading/trailing newlines
+        ("\r\n mixed \n newlines \r\n", " mixed  newlines "),  # Mixed with spaces
+    ],
+)
+def test_injection_multiline_secret_stripped(repo_tmp, secret_input, expected_injected):
```
[the remark this answers is not in the document]
Comment (ikostan) [reply tests/ci/test_salt_injection.py:263]: I think this one is fixed. Please check.

[the remark this answers is not in the document]
Comment (ikostan) [reply tests/ci/test_salt_injection.py:263]: Please check again.

File: tests/ci/test_salt_injection.py [hunk 7]
```diff
@@ -252,3 +232,38 @@ def test_idempotency(repo_tmp):
...
+    """
+    Validates that multiline secrets (both LF and CRLF) are safely stripped
+    of all newlines and injected, rather than failing or corrupting the file.
+    """
+    dummy_rel = f"{repo_tmp}/dummy_multiline_{hash(secret_input)}.gd"
+    dummy_abs = Path(PROJECT_ROOT) / dummy_rel
+    original_content = 'var salt = "CI_INJECT_SALT_HERE"\n'
+
+    dummy_abs.write_text(original_content, encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (ikostan) [reply tests/ci/test_salt_injection.py:272]: Here is exactly what the bot is talking about:

Right now, we tests pass an empty string (""). But in Bash, an empty string is still technically "set." Because we added set -euo pipefail to the bash script, if the GitHub environment variable completely fails to load (making it truly unset), the -u flag wants to instantly crash the script with an ugly unbound variable error.

[the remark this answers is not in the document]
Comment (ikostan) [reply tests/ci/test_salt_injection.py:272]: Did I fix it?

File: .github/scripts/inject_salt.sh [hunk 8]
```diff
@@ -13,22 +17,38 @@ if [ ! -f "$TARGET_FILE" ]; then
...
+fi
+
 # Source the shared utilities
 source "$(dirname "$0")/ci_utils.sh"
 
 # 3. Replace the safe placeholder with the real secret
 sedi "s|\"CI_INJECT_SALT_HERE\"|\"$SED_ESCAPED\"|g" "$TARGET_FILE"
+
+# 🛑 4. EXPLICIT VERIFICATION: Ensure the placeholder was actually removed
+if grep -q "CI_INJECT_SALT_HERE" -- "$TARGET_FILE"; then
```
[the remark this answers is not in the document]
Comment (ikostan) [reply .github/scripts/inject_salt.sh]: The bot actually caught a syntax mistake on my part. I completely misplaced the -- separator!

File: .github/scripts/inject_salt.sh [hunk 9]
```diff
@@ -13,22 +17,38 @@ if [ ! -f "$TARGET_FILE" ]; then
...
+CLEAN_SALT=$(printf '%s' "$PRODUCTION_SALT" | tr -d '\r\n')
+
 echo "⚙️ Injecting secret into $TARGET_FILE..."
 
 # 1. Escape for Godot's parser
-GODOT_ESCAPED=$(printf '%s' "$PRODUCTION_SALT" | sed 's/\\/\\\\/g; s/"/\\"/g')
+GODOT_ESCAPED=$(printf '%s' "$CLEAN_SALT" | sed 's/\\/\\\\/g; s/"/\\"/g')
 
 # 2. Escape for sed replacement string
 SED_ESCAPED=$(printf '%s' "$GODOT_ESCAPED" | sed 's/\\/\\\\/g; s/&/\\&/g; s/|/\\|/g')
```
[the remark this answers is not in the document]
Comment (ikostan) [reply .github/scripts/inject_salt.sh:35]: Our production secrets are always single-line high-entropy strings, squashing all newlines globally is completely fine. It guarantees that no matter how badly a copy-paste messes up a GitHub Secret, it gets squashed into a clean string.

**Comment by ikostan** [comment]
> * In `deploy_to_itch.yml`, the unzip destination `export/patched_build` is hard-coded while the source/target zip path uses `${{ steps.export.outputs.archive_directory }}`; aligning the patched folder path with that output (e.g., deriving the patched path from `archive_directory`) would reduce coupling to the current `export/...` layout and make the workflow more robust to path changes.

Addressed

**Comment by ikostan** [comment]
> * In `test_injection_multiline_secret_stripped`, using `hash(secret_input)` to build file names ties the test data to Python’s randomized hash (and may yield awkward names with `-`); consider using a deterministic/sanitized identifier (e.g., an index from `enumerate` or a small helper that strips/normalizes characters) for clearer, stable filenames.

Working on it.

File: .github/scripts/inject_salt.sh [hunk 10]
```diff
@@ -13,22 +17,38 @@ if [ ! -f "$TARGET_FILE" ]; then
...
+
 # Source the shared utilities
 source "$(dirname "$0")/ci_utils.sh"
 
 # 3. Replace the safe placeholder with the real secret
 sedi "s|\"CI_INJECT_SALT_HERE\"|\"$SED_ESCAPED\"|g" "$TARGET_FILE"
+
+# 🛑 4. EXPLICIT VERIFICATION: Ensure the placeholder was actually removed
+if grep -q -- "CI_INJECT_SALT_HERE" "$TARGET_FILE"; then
+    echo "❌ FATAL: Salt injection failed! Placeholder still exists in $TARGET_FILE."
```
[the remark this answers is not in the document]
Comment (ikostan) [reply .github/scripts/inject_salt.sh]: Fixed

File: tests/ci/test_salt_injection.py [hunk 11]
```diff
@@ -252,3 +238,64 @@ def test_idempotency(repo_tmp):
...
+@pytest.mark.parametrize(
+    "secret_input, expected_injected",
+    [
+        ("line1\nline2", "line1line2"),  # Standard Linux/macOS LF
+        ("line1\r\nline2", "line1line2"),  # Windows CRLF
+        ("\n\nline1\n\nline2\n\n", "line1line2"),  # Multiple leading/trailing newlines
+        ("\r\n mixed \n newlines \r\n", " mixed  newlines "),  # Mixed with spaces
+        ("   ", "   "),  # All-whitespace (valid string, tests positive path)
+        ("  \n\n  ", "    "),  # Whitespace split by newlines
+    ],
```
[the remark this answers is not in the document]
Comment (ikostan) [reply tests/ci/test_salt_injection.py:261]: Yes, we absolutely addressed that one!

The Sourcery bot is complaining because it wants proof that a secret consisting *only* of newlines (which becomes completely empty after the `tr -d '\r\n'` step) will properly fail the pipeline and leave the target file untouched.

If we look at the **`test_salt_injection.py`** file, right around line 144, we can see we already nailed this exact requirement:

```python
@pytest.mark.parametrize(
    "empty_input",
    [
        "",  # Completely empty
        "\n",  # Single Linux newline
        "\r\n\r\n",  # Multiple Windows newlines
    ],
)
def test_injection_empty_secret(repo_tmp, empty_input):
```

Inside that test, we explicitly assert the exact two things the bot requested:

1. The script aborts gracefully (`assert result.returncode != 0`).
2. The file is left completely unchanged (`assert dummy_abs.read_text(encoding="utf-8") == original_content`).

Because we split this out into its own dedicated negative-path test (`test_injection_empty_secret`) rather than stuffing it into the positive-path multiline test, the bot might have gotten confused reading the diff.

We are 100% covered for this edge case. We can confidently hit "Resolve conversation" on that thread!

**Comment by ikostan** [comment]
> * In `run_injection`, you set `env['PRODUCTION_SALT']` before the `None` check and then conditionally override/pop it—consider removing the initial unconditional assignment to avoid redundant writes and make the control flow clearer.

Working on it...

**Comment by ikostan** [comment]
> * In `inject_salt.sh`, the numeric comment labels (`2.`, `2.5.` etc.) can become misleading if the script evolves; consider simplifying these to descriptive headings without implied ordering to reduce future maintenance friction.

Working on it...

**Comment by ikostan** [comment]
> * In the Itch deployment workflow, the re-zip step uses a relative path with multiple `../` segments; using `pushd/popd` or constructing absolute paths from `${{ steps.export.outputs.archive_directory }}` would make the archive path handling more robust and easier to reason about.

We improved this earlier by dropping into the dynamic folder, but zip -r "../Web.zip" . still relies on relative traversal. By using pushd/popd and GitHub's built-in $GITHUB_WORKSPACE variable, we can make the destination path absolute and completely bulletproof.

**Comment by ikostan** [comment]
@sourcery-ai review

**Review by SourceryAI — COMMENTED** [review]
Hey - I've left some high level feedback:

- The new `log_message("Generated key length..."` in `_get_encryption_key()` will run on every key generation in production; consider gating this behind a debug flag or a `OS.has_feature("ci")`/dev build check to avoid noisy logs in normal gameplay.
- In the `Verify export does not contain placeholder` workflow step you disable `pipefail` but never restore it; if more commands are added later in this step, it would be safer to capture the previous `set -o` state or re-enable `pipefail` after the `grep` check.

<details>
<summary>Prompt for AI Agents</summary>

~~~markdown
Please address the comments from this code review:

## Overall Comments
- The new `log_message("Generated key length..."` in `_get_encryption_key()` will run on every key generation in production; consider gating this behind a debug flag or a `OS.has_feature("ci")`/dev build check to avoid noisy logs in normal gameplay.
- In the `Verify export does not contain placeholder` workflow step you disable `pipefail` but never restore it; if more commands are added later in this step, it would be safer to capture the previous `set -o` state or re-enable `pipefail` after the `grep` check.
~~~

</details>

***

Hi @ikostan! 👋

Thanks for trying out Sourcery by commenting with `@sourcery-ai review`! 🚀

Install the [sourcery-ai bot](https://github.com/apps/sourcery-ai/installations/new) to get automatic code reviews on every pull request ✨

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>

File: .github/workflows/deploy_to_itch.yml [hunk 12]
```diff
@@ -64,16 +70,45 @@ jobs:
           use_preset_export_path: true
 
       - name: "Unzip Export Archive"
-        run: unzip export/web/Web.zip -d export/web
+        run: |
+          set -euo pipefail
+          # Defensive hygiene: ensure clean staging directory before unzipping
+          rm -rf "${{ steps.export.outputs.archive_directory }}/patched_build"
+          mkdir -p "${{ steps.export.outputs.archive_directory }}/patched_build"
+          unzip "${{ steps.export.outputs.archive_directory }}/Web.zip" -d "${{ steps.export.outputs.archive_directory }}/patched_build"
```
[the remark this answers is not in the document]
Comment (ikostan) [reply .github/workflows/deploy_to_itch.yml]: Sourcery actually caught a legendary DevOps trap here!

The bot is absolutely right. Using rm -rf "$VARIABLE/some_folder" is a classic footgun in bash scripting. If the Godot export step silently fails and the archive_directory variable ends up being completely empty, that line evaluates to rm -rf "/patched_build". It will literally try to delete a folder from the root directory of the Linux runner.

To make this step mathematically immune to that failure, we just need to assign the variable explicitly and throw a fatal error if it's empty, /, or ..

File: scripts/core/globals.gd [hunk 13]
```diff
@@ -488,7 +488,13 @@ func _get_encryption_key() -> String:
...
 		device_id = OS.get_unique_id()
 
-	return (device_id + salt).sha256_text()
+	# Logging the key length immediately after generation gives you the cleanest
+	# signal whether the real production salt was used.
+	var final_key: String = (device_id + salt).sha256_text()
+	# NEW: Log key length for easier debugging of injection issues
+	log_message("Generated key length: %d" % final_key.length(), LogLevel.DEBUG)
+
+	return final_key
```
[the remark this answers is not in the document]
Comment (ikostan) [reply scripts/core/globals.gd:497]: Sourcery is completely missing the forest for the trees here. You can safely dismiss this suggestion!

The bot is making two fundamentally flawed assumptions about how your specific architecture works:

1. **The Frequency Hallucination:** The bot claims this will create "a lot of log noise" because it runs on every call. But `_get_encryption_key()` is not a hot-loop function. It is called exactly *once* when the singleton initializes the `save_encryption_pass` variable at the top of the script, and maybe one additional time as a failsafe inside `ensure_encryption_key()`. It will print to the console exactly once per game boot.
2. **Ignoring Our Custom Logger:** The bot wants we to hardcode an `OS.is_debug_build()` check. But we already built a robust custom logging architecture to handle this exact scenario! Because we are passing `LogLevel.DEBUG` to our `log_message` function, this log is *already* silenced in production builds unless the user explicitly toggles `settings.enable_debug_logging` to true.

Adding an OS-level debug check here would just add redundant boilerplate and bypass your own custom settings infrastructure. We are fully covered—hit "Dismiss" on that comment!

**Comment by ikostan** [comment]
> * The new `log_message("Generated key length..."` in `_get_encryption_key()` will run on every key generation in production; consider gating this behind a debug flag or a `OS.has_feature("ci")`/dev build check to avoid noisy logs in normal gameplay.

Won't fix

**Comment by ikostan** [comment]
> * In the `Verify export does not contain placeholder` workflow step you disable `pipefail` but never restore it; if more commands are added later in this step, it would be safer to capture the previous `set -o` state or re-enable `pipefail` after the `grep` check.

You are 100% correct. That is an incredibly sharp catch and a textbook example of defensive bash scripting.

Modifying shell state (`set +o`) and leaving it dangling is a trap waiting to be sprung. If someone comes along in six months and appends another pipeline command to the bottom of that step, it would silently operate without `pipefail` protection, completely undermining the strict mode we just instituted.

Here is the properly scoped update for that specific step in **`.github/workflows/deploy_to_itch.yml`** to ensure the state is restored immediately after the hazardous pipeline executes:

```yaml
      - name: "Verify export does not contain placeholder"
        run: |
          set -euo pipefail
          
          # Temporarily disable pipefail to prevent SIGPIPE from failing the check 
          # if grep exits early upon finding a match in the unpacked byte stream.
          set +o pipefail 
          
          if unzip -p "${{ steps.export.outputs.archive_directory }}/Web.zip" | grep -qF "CI_INJECT_SALT_HERE"; then
            echo "❌ Placeholder detected inside final export archive."
            exit 1
          fi
          
          # Immediately restore strict pipefail for any future commands added to this block
          set -o pipefail
          
          echo "✅ Final export archive verified successfully."
```

By re-enabling `set -o pipefail` right after the `if` block, the mutation is perfectly encapsulated. The step is now completely future-proofed against subsequent edits. Brilliant review!
